indico/modules/events/timetable/controllers/display.py

```diff
@@ -73,28 +80,64 @@ def _process(self):
+
         if form.validate_on_submit():
-            form_data = form.data_for_format
-            pdf_format = TimetablePDFFormat(form_data)
-            if not form.advanced.data:
-                pdf_format.contribsAtConfLevel = True
-                pdf_format.breaksAtConfLevel = True
-                pdf_class = SimplifiedTimeTablePlain
-                additional_params = {}
-            else:
-                pdf_class = TimeTablePlain
-                additional_params = {'firstPageNumber': form.firstPageNumber.data,
-                                     'showSpeakerAffiliation': form_data['showSpeakerAffiliation'],
-                                     'showSessionDescription': form_data['showSessionDescription']}
-            if request.args.get('download') == '1':
-                pdf = pdf_class(self.event, session.user, sortingCrit=None, ttPDFFormat=pdf_format,
-                                pagesize=form.pagesize.data, **additional_params)
-                return send_file('timetable.pdf', BytesIO(pdf.getPDFBin()), 'application/pdf')
+            days = {}
+            now = now_utc()
+            css = render_template('events/timetable/pdf/timetable.css')
+
+            for day in self.event.iter_days():
+                entries = (self.event.timetable_entries
+                           .filter(
+                               db.cast(TimetableEntry.start_dt.astimezone(self.event.tzinfo), db.Date) == day,
+                               TimetableEntry.parent_id.is_(None))
+                           .order_by(TimetableEntry.start_dt))
+                days[day] = entries
+
+            config = TimetableExportConfig(
+                show_title=form.other.data['showSpeakerTitle'],
+                show_affiliation=form.other.data['showSpeakerAffiliation'],
+                show_cover_page=form.document_settings.data['showCoverPage'],
+                show_toc=form.document_settings.data['showTableContents'],
+                show_session_toc=form.document_settings.data['showSessionTOC'],
+                show_abstract=form.contribution_info.data['showAbstract'],
+                dont_show_poster_abstract=form.contribution_info.data['dontShowPosterAbstract'],
+                show_contribs=form.visible_entries.data['showContribsAtConfLevel'],
+                show_length_contribs=form.contribution_info.data['showLengthContribs'],
+                show_breaks=form.visible_entries.data['showBreaksAtConfLevel'],
+                new_page_per_session=form.session_info.data['newPagePerSession'],
+                show_session_description=form.session_info.data['showSessionDescription'],
+                print_date_close_to_sessions=form.session_info.data['printDateCloseToSessions']
+            )
+
+            html = render_template('events/timetable/pdf/timetable.html', event=self.event,
+                                   days=days, now=now, config=config)
```

Comment:
This and all the querying is not needed outside `download` mode, so right now the (expensive) query (or rather queries atm) are executed twice, even though the first time they're completely useless.

Reply:
Fixed by moving the download check to the top as such:

```python
        if form.validate_on_submit():
            if not download:
                url = url_for(request.endpoint, **dict(request.view_args, download=True, **request.args.to_dict(False)))
                return jsonify_data(flash=False, redirect=url, redirect_no_loading=True)             
```